### PR TITLE
Create and upload a test coverage report to coveralls

### PR DESCRIPTION
**Pull request by superm1**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: data/installed-tests/fwupdtool.sh [hunk 1]
```diff
@@ -9,10 +9,15 @@ INPUT="@installedtestsdir@/fakedevice124.bin \
 
 # ---
 echo "Building ${CAB}..."
+rm -f ${CAB}
 fwupdtool build-cabinet ${CAB} ${INPUT}
```
Comment (hughsie) [inline data/installed-tests/fwupdtool.sh:13]: this feels like we should teach build-cabinet about `--force` to allow overwriting.

**Comment by hughsie** [comment]
Conceptually I think this is a good idea. Are you thinking of making it gate new PRs?

**Comment by superm1** [comment]
> Conceptually I think this is a good idea. Are you thinking of making it gate new PRs?

I don't know yet.  I want to see how the numbers actually are when all the tests have _actually_ run.  I need to fight with the CI jobs a bunch for that though.

**Comment by superm1** [comment]
Even though it's running end to end something still seems really wrong - I think it's missing a lot of files still.

**Comment by superm1** [comment]
Huh; even combining jobs (to keep the workspace) it's missing a ton.  Any ideas?

**Comment by hughsie** [comment]
> ► 86.76 fwupdplugin-self-test.p/

It's only got the *generated* files? e.g. https://coveralls.io/builds/65530463

**Comment by superm1** [comment]
> > ► 86.76 fwupdplugin-self-test.p/
> 
> It's only got the _generated_ files? e.g. https://coveralls.io/builds/65530463

Yeah; I can't figure out why though.  I wonder if some gcovr stuff gets stripped somehow in the Debian packaging process.

I'd push this out to a different set of jobs, but we need one of them that runs **all** the tests and compiles everything.

**Comment by hughsie** [comment]
> I wonder if some gcovr stuff gets stripped somehow

Is the debian package explicitly running `strip` on the binaries? I think we need gcov pre-package.

**Comment by superm1** [comment]
> > I wonder if some gcovr stuff gets stripped somehow
> 
> Is the debian package explicitly running `strip` on the binaries? I think we need gcov pre-package.

That was my conclusion too - stripping the binaries is what's causing it.  But I just pushed some changes to avoid stripping binaries and it's still happening...

**Comment by superm1** [comment]
I downloaded and installed the artifacts from [ad1889e](https://github.com/fwupd/fwupd/pull/6777/commits/ad1889e774af7f75d7fb267b573a270ea95d7011) and just tried to run fwupdtool on a container.  I see that this entire directory structure gets built when I run `fwupdtool`

`/github/workspace/build/obj-x86_64-linux-gnu/libfwupdplugin/libfwupdplugin.so.p/`:

```
fu-acpi-table.c.gcda              fu-coswid-firmware.c.gcda             fu-efi-section.c.gcda         fu-hwids-kenv.c.gcda                  fu-path.c.gcda                                  meson-generated_.._fu-cab-struct.c.gcda
fu-archive.c.gcda                 fu-crc.c.gcda                         fu-efi-signature.c.gcda       fu-hwids-smbios.c.gcda                fu-pefile-firmware.c.gcda                       meson-generated_.._fu-cfu-firmware-struct.c.gcda
fu-archive-firmware.c.gcda        fu-csv-entry.c.gcda                   fu-efi-signature-list.c.gcda  fu-i2c-device.c.gcda                  fu-plugin.c.gcda                                meson-generated_.._fu-coswid-struct.c.gcda
fu-backend.c.gcda                 fu-csv-firmware.c.gcda                fu-efivar.c.gcda              fu-ifd-bios.c.gcda                    fu-progress.c.gcda                              meson-generated_.._fu-dfu-firmware-struct.c.gcda
fu-bios-setting.c.gcda            fu-device.c.gcda                      fu-efivar-linux.c.gcda        fu-ifd-common.c.gcda                  fu-quirks.c.gcda                                meson-generated_.._fu-dpaux-struct.c.gcda
fu-bios-settings.c.gcda           fu-device-locker.c.gcda               fu-efi-volume.c.gcda          fu-ifd-firmware.c.gcda                fu-sbatlevel-section.c.gcda                     meson-generated_.._fu-edid-struct.c.gcda
fu-bluez-device.c.gcda            fu-device-progress.c.gcda             fu-elf-firmware.c.gcda        fu-ifd-image.c.gcda                   fu-security-attr.c.gcda                         meson-generated_.._fu-efi-struct.c.gcda
fu-byte-array.c.gcda              fu-dfu-firmware.c.gcda                fu-fdt-firmware.c.gcda        fu-ifwi-cpd-firmware.c.gcda           fu-security-attrs.c.gcda                        meson-generated_.._fu-elf-struct.c.gcda
fu-bytes.c.gcda                   fu-dfuse-firmware.c.gcda              fu-fdt-image.c.gcda           fu-ifwi-fpt-firmware.c.gcda           fu-smbios.c.gcda                                meson-generated_.._fu-fdt-struct.c.gcda
fu-cab-firmware.c.gcda            fu-dpaux-device.c.gcda                fu-firmware.c.gcda            fu-ihex-firmware.c.gcda               fu-srec-firmware.c.gcda                         meson-generated_.._fu-fmap-struct.c.gcda
fu-cab-image.c.gcda               fu-drm-device.c.gcda                  fu-firmware-common.c.gcda     fu-input-stream.c.gcda                fu-string.c.gcda                                meson-generated_.._fu-hid-struct.c.gcda
fu-cfi-device.c.gcda              fu-dump.c.gcda                        fu-fit-firmware.c.gcda        fu-intel-thunderbolt-firmware.c.gcda  fu-sum.c.gcda                                   meson-generated_.._fu-ifd-struct.c.gcda
fu-cfu-offer.c.gcda               fu-edid.c.gcda                        fu-fmap-firmware.c.gcda       fu-intel-thunderbolt-nvm.c.gcda       fu-udev-device.c.gcda                           meson-generated_.._fu-ifwi-struct.c.gcda
fu-cfu-payload.c.gcda             fu-efi-common.c.gcda                  fu-hid-descriptor.c.gcda      fu-io-channel.c.gcda                  fu-usb-device.c.gcda                            meson-generated_.._fu-intel-thunderbolt-struct.c.gcda
fu-chunk-array.c.gcda             fu-efi-device-path.c.gcda             fu-hid-device.c.gcda          fu-kenv.c.gcda                        fu-usb-device-ds20.c.gcda                       meson-generated_.._fu-oprom-struct.c.gcda
fu-chunk.c.gcda                   fu-efi-device-path-list.c.gcda        fu-hid-report.c.gcda          fu-kernel.c.gcda                      fu-usb-device-fw-ds20.c.gcda                    meson-generated_.._fu-pefile-struct.c.gcda
fu-common.c.gcda                  fu-efi-file.c.gcda                    fu-hid-report-item.c.gcda     fu-linear-firmware.c.gcda             fu-usb-device-ms-ds20.c.gcda                    meson-generated_.._fu-sbatlevel-section-struct.c.gcda
fu-common-guid.c.gcda             fu-efi-file-path-device-path.c.gcda   fu-hwids.c.gcda               fu-lzma-common.c.gcda                 fu-uswid-firmware.c.gcda                        meson-generated_.._fu-smbios-struct.c.gcda
fu-common-linux.c.gcda            fu-efi-filesystem.c.gcda              fu-hwids-config.c.gcda        fu-mei-device.c.gcda                  fu-version-common.c.gcda                        meson-generated_.._fu-usb-device-ds20-struct.c.gcda
fu-composite-input-stream.c.gcda  fu-efi-hard-drive-device-path.c.gcda  fu-hwids-darwin.c.gcda        fu-mem.c.gcda                         fu-volume.c.gcda                                meson-generated_.._fu-uswid-struct.c.gcda
fu-config.c.gcda                  fu-efi-load-option.c.gcda             fu-hwids-dmi.c.gcda           fu-oprom-firmware.c.gcda              meson-generated_.._fu-acpi-table-struct.c.gcda
fu-context.c.gcda                 fu-efi-lz77-decompressor.c.gcda       fu-hwids-fdt.c.gcda           fu-partial-input-stream.c.gcda        meson-generated_.._fu-archive-struct.c.gcda
```

So it seems like it's generating coverage information, maybe the paths are wrong where it's all run?

**Comment by superm1** [comment]
So all the `gcno` and `gcda` files are there, and `gcovr` report actually matches everything too.  Seems like maybe something goes wrong with the _meson_ **coverage** target not working.  Will capture artifacts from that next I guess.

**Comment by superm1** [comment]
Yup; so confirming that the coverage target is missing data, but manually running `gcovr` works.

**Comment by superm1** [comment]
>  First build on mlimonci/tests at 38.317%

Ah my theory is right.  It's working now!  Let me squash and minimize what's really needed.

**Comment by superm1** [comment]
Alright; it's all working for real now.  With the number we have I think we should merge this, but we shouldn't gate anything on it until our baseline threshold is higher.

**Comment by hughsie** [comment]
@superm1 why are there *3* different coverage scores? Do we also want to exclude obj-x86_64-linux-gnu? it's only generated code as far as I can see, and clicking on any of the results I see `The file "obj-x86_64-linux-gnu/libfwupdplugin/fu-pefile-struct.c" isn't available on github. Either it's been removed, or the repo root directory needs to be updated.
`

**Comment by hughsie** [comment]
SO maybe we don't want to exclude it, but *fix* it instead. e.g. https://coveralls.io/builds/65541348/source?filename=obj-x86_64-linux-gnu%2Fplugins%2Facpi-phat%2Flibfu_plugin_acpi_phat.a.p%2Ffu-acpi-phat-struct.c has 0% coverage, but we must be compiling and using that in the self tests -- no?

**Comment by superm1** [comment]
> SO maybe we don't want to exclude it, but fix it instead. e.g. https://coveralls.io/builds/65541348/source?filename=obj-x86_64-linux-gnu%2Fplugins%2Facpi-phat%2Flibfu_plugin_acpi_phat.a.p%2Ffu-acpi-phat-struct.c has 0% coverage, but we must be compiling and using that in the self tests -- no?

Maybe a bunch of funcs declared but not really used in self tests.  Compare to dfu for example:

https://coveralls.io/builds/65541348/source?filename=obj-i686-linux-gnu%2Fplugins%2Fdfu%2Ffu-dfu-struct.c

**Review by hughsie — APPROVED** [review]
So maybe we just have to make the numbers go up by testing stuff in the self tests rather than filtering out the bad results :)

**Comment by superm1** [comment]
> So maybe we just have to make the numbers go up by testing stuff in the self tests rather than filtering out the bad results :)

Exactly!  We should at least aim for the number to not drop in the future.

**Pull request merged**